tests/mir_test_framework/window_management_test_harness.cpp

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mir/test/doubles/fake_display.h"
+#include "mir/test/doubles/stub_buffer_stream.h"
+#include "mir_test_framework/window_management_test_harness.h"
+#include <mir/executor.h>
+#include <mir/scene/surface.h>
+#include <mir/scene/surface_observer.h>
+#include <mir/shell/surface_specification.h>
+#include <mir/shell/shell.h>
+#include <mir/wayland/weak.h>
+#include <miral/window_manager_tools.h>
+#include <miral/set_window_management_policy.h>
+
+#include <queue>
+
+namespace ms = mir::scene;
+namespace msh = mir::shell;
+namespace mf = mir::frontend;
+namespace mi = mir::input;
+namespace mtd = mir::test::doubles;
+namespace mg = mir::graphics;
+namespace mc = mir::compositor;
+namespace geom = mir::geometry;
+using namespace testing;
+
+namespace
+{
+class WorkQueue
+{
+public:
+    void enqueue(std::function<void()> const& f)
+    {
+        work_queue.push(f);
+    }
+
+    auto next() -> bool
+    {
+        if (work_queue.empty())
+            return false;
+
+        auto const& f = work_queue.front();
+        f();
+        work_queue.pop();
```

Comment:
This can result in infinite recursion, right? Almost all usages of the workqueue are of the form `queue.enqueue(...); queue.run()`, and if that gets run *from* the workqueue it'll push something to the back of the queue and then run itself again?

Reply:
Yeah I agree that it's kind of nasty... I am trying to find a good way to avoid both the nastiness of the queue and the nastiness of asynchronicity

Reply:
After thinking about this deeper, it _can_ result in recursion, but this isn't necessarily a bad thing. The `WorkQueue` will just continually be run recursively, which will work.

If you end up in a situation where the recursion is _infinite_, then that suggests that your WM is doing something wrong. e.g. if every time you resize a window, you also move it, and every time you move a window, you also resize it, then that is on you

Reply:
So, the problem I'm talking about is running `f()` *before* running `work_queue.pop()`. If `f()` happens to run `work_queue.run()`, then `work_queue.run()` will call `work_queue.front()`, receive *the same* `f` again, and recurse.

If you popped before calling `f()` (and `std::move()`d out of `front()`), that wouldn't happen.